api/swagger.yaml

```diff
@@ -10095,7 +10095,7 @@ paths:
           schema:
             $ref: "#/definitions/ErrorResponse"
         403:
-          description: "Operation not supported for swarm scoped networks"
+          description: "Operation not supported for swarm scoped networks. Also happens when the network is already connected and the container is currently running"
```

Comment:
Thanks for contributing!

I wonder if we should make the description more generic; I'd like to avoid trying to describe every possible error-condition, as more may come, and we won't be able to describe every possible condition for errors to occur, unless we create specific error-codes for each (we have discussed adding a custom error-code field to the response for specific cases, but we currently don't)

I had a quick peek at history, and I see I added these originally in https://github.com/moby/moby/commit/b0089e48272f18d856ba147b393371c18d5683fb (https://github.com/moby/moby/pull/25002). We didn't have a swagger file yet at that time, and the API docs were a bit more free-form. In hindsight, it would've been better to keep the error-description for the 403 status _itself_ more generic; at the time this may have been the only condition that returned a 403, but more conditions may occur, so a 403 should likely not be considered "exclusive" for this case.

Maybe instead;

- make the error-description more generic, keeping the error-message in the swagger purely as example (the actual error message itself should provide details about the reason something happened, but none of that is part of the API's contract currently)
- _Perhaps_ (as the swarm case is a more "expected" condition), we could instead describe that as part of the endpoint's description (there are some other endpoints which also return a 403 for the Swarm case).


Just a quick draft (better suggestions definitely welcome!) for the endpoint's description (alternatively, we could reverse the wording and describe the error-condition; not sure if that's clearer though);

```
Connect a container to a network.

The container must be running, and the network must be either a local-scoped
network or a swarm-scoped network with the "attachable" option set.
```

For the error, we could keep it at a generic "Operation forbidden"

Let me know what you think!

Reply:
Thinking about this; I'm also wondering if the `403` for the container not running case is correct. wondering if a more generic 400, or a [409 (conflict)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409), or even a [412](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412) would make more sense 🤔

Reply:
Thanks for the suggestions! I updated the description and the error message. This error happens when the container is already running and the connect endpoint is called again (the first connect works flawlessly).

I am not sure about the specific status code, in my opinion something like 200 or 204 would also be fine since the operation is basically a no-op. The network has already been connected to the container previously.

However, I would like to avoid changing the status code now because that might brake projects that already depend on this code in their application logic like the aforementioned nextcloud-aio

Reply:
> This error happens when the container is already running and the connect endpoint is called again (the first connect works flawlessly).

Ah! Yes, that one makes sense. I was actually discussing this PR with @akerouanton, and while discussing, he mentioned "but connecting to a non-running container should work", so if that didn't work with PodMan, that would be a potential bug in that implementation;

```bash
docker create --name mycontainer busybox
docker network create mynetwork
docker network connect mynetwork mycontainer
```

> in my opinion something like 200 or 204 would also be fine since the operation is basically a no-op. The network has already been connected to the container previously.

I recall I discussed some of that in the past, and if I recall correctly, the main reason was because `network connect` also accepts options, so in case those options differ between multiple connects, they should be compared with the existing connection. Silently ignoring duplicates there could override earlier options (and possibly not all options could be modified on an existing connection).

So while it may be possible to account for that, this may also require some additional work (which hasn't been done (yet)). In the meantime, producing an error means that we keep that option open for future (because nobody can currently depend on that feature, as it's currently an error-condition).